perceval/converters/circuit_to_graph_converter.py

```diff
@@ -0,0 +1,188 @@
+                for j in range(i + 1, len(subset)):
+                    edge = (subset[i], subset[j])
+                    # Increase the weight of the edge by 1 if the gate is in the list. Increase by 2 if not.
+                    if self.gates[l] in ['cx', 'cy', 'cz', 'ch', 'cp']:
+                        if edge in edge_weights:
+                            edge_weights[edge] += 1
+                        else:
+                            edge_weights[edge] = 1
+                    elif self.gates[l] in ['swap', 'iswap']:
+                        # Free operations.
+                        pass
+                    else:
+                        if edge in edge_weights:
+                            edge_weights[edge] += 2
+                        else:
+                            edge_weights[edge] = 2
+
+        for edge, weight in edge_weights.items():
+            g.add_edge(*edge, weight=weight)
+        return g
+
+    @staticmethod
+    def plot_graph(g):  # g is the graph created in graph_generator
+        pos = nx.spring_layout(g, seed=42)
+        nx.draw_networkx_nodes(g, pos, node_size=90, node_color='b')
+        nx.draw_networkx_edges(g, pos)
+        nx.draw_networkx_labels(g, pos, font_size=10, font_color='white', font_family="sans-serif")
+        edge_labels = nx.get_edge_attributes(g, "weight")
+        nx.draw_networkx_edge_labels(g, pos, edge_labels)
+        plt.show()
+
+    def graph_k_clustering_and_cnots_needed(self,
+                                            compute_with_min_cnots: bool = False) -> tuple[list[list[int]], list[int]]:
+        """''
+        Computes the laplacian matrix of the graph, compute its eigenvectors sorted by their
+        eigenvalues. For all the possible subpartitions of the graph, it will compute the
+        kmeans clustering with the respective number of eigenvectors as features.
+        Given the random initial state for the kmeans method, the user has the option to compute
+        multiple repetitions and choose the partition that gives the smallest number of CNOTs.
+
+        :param compute_with_min_cnots: Whether to compute it 30 times and choose the output yielding to the minimum of CNOTs.
+        :type compute_with_min_cnots: bool
+        :return: Clustering result and CNOT counts.
+        :rtype: tuple[list[list[int]], list[int]]
+        '"""
+        graph = self.graph_generator()
+        laplacian_matrix = nx.normalized_laplacian_matrix(graph).toarray()
+        eigenvalues, eigenvectors = np.linalg.eigh(laplacian_matrix)
+        n = len(graph.nodes)  # number of qubits
+        possible_partitions = []
+        number_of_cnots = []
+
+        for k in range(2, n + 1):
+            # 1 is the trivial case of the complete circuit in one qudit and n is the dual rail.
+            features = eigenvectors[:, :k]
+            if compute_with_min_cnots:
+                best_partition = None
+                min_cnots = float('inf')
+                for _ in range(30):
```

Comment:
This hard-coded value of `30` should be replaced by variable name to increase readability of the code. If it is some fixed number, I suggest adding `CAPTILIZED_VARIABLE_NAME=30` at the beginning of the file (outside class scope) and then using it here.